### ISSUE TITLE
Improve artifact / image management

### ISSUE THREAD

**Issue by lmars** [flynn/flynn#2422]
Here are some proposed changes for improving image management and artifact creation:
- Switch to using OverlayFS rather than AUFS
  
  Given OverlayFS is now in the mainline kernel, other than having a minimum requirement on kernel version we would not need to install extra packages [as we do for AUFS](https://github.com/flynn/flynn/blob/master/script/install-flynn.tmpl#L102-L112). OverlayFS is also simpler than AUFS and is reported to be faster.
- Support creating artifacts from a job's filesystem changes
  
  `flynn-host` creates a layer for storing the filesystem changes made by a job, but this is temporary and removed when the job stops. We should support keeping this layer around so that it can be fetched, stored in the blobstore and an artifact created for it. We would likely want to support include / exclude rules so we could run a build job but only save the result of the build and not any intermediary files.
- Support multiple artifacts per release
  
  Rather than jobs having a single image pulled from the release's only artifact, we should support releases having an ordered list of artifacts which are overlayed under the job's layer. We could then migrate slugs from being tarball URLs stored in environment variables to artifacts which are overlayed over the slugrunner artifact (meaning slugs wouldn't be fetched / extracted every time a job starts, they would be cached in the host's image storage). This would also give us the flexibility to have arbitrary build pipelines which create artifacts and make them available to jobs (e.g. have a Ruby artifact which can be added to a release if a job needs Ruby), moving us closer to supporting #1865.

Thoughts very much welcome on all of the above, but my particular concerns:
- storing layers in the blobstore means they are stored in postgres, and so are replicated across the cluster, but then when `flynn-host` pulls the layer, it will also have it stored, so we are in effect storing the layer twice on some hosts. It would be better if `flynn-host` itself was the service from which layers were pulled (so an artifact URL would be something like `http://leader.flynn-host.discoverd/image/<id>`) but then we need to think about HA / replication at that level, is that something we want?
- overlaying artifacts means higher layers can "overwrite" files from lower layers, which could cause incompatibilities (e.g. if the base layer has `libfoo` symlinked to `libfoo.2.0` and a later layer changes that symlink to `libfoo.3.0`, library loading may break). I guess we need to make this clear (and maybe convince people to use [Nix](http://nixos.org/nix/) :wink:)

**Comment by josephglanville** [comment]
Few thoughts.

* If we are going to move to OverlayFS we also need to look at security. I don't know how well it plays with user namespaces for example which is really important for being able to give containers root.

* Same goes for examining ZFS for this usecase instead. As it has some desirable properties for it's use as a CoW filesystem in a cluster setting where the ability to stream the changes in send/receive format (which is both more efficient and uses less IO to apply). It would also drop the dependency on a super recent kernel version which I think isn't the biggest deal but may prove handy.
ZFS is also just a really good filesystem that has been in much wider use than OverlayFS for a much longer time.

* Storing the artifacts/layers in the blobstore should be ok, the blobstore is more of a gateway interface anyway. Long term it seems reasonable that we will switch to backing it with a "real" blob storage system designed to operate with lower overheads than PG replication (namely use of erasure coding I would assume). Though it is actually possible that we may be able to use ZFS for this instead with some cute hacks to achieve replication etc. It will still have similar overheads to the PG option but probably much better performance.

**Comment by titanous** [comment]
- We should keep user images in blobstore, having multiple copies around isn't really a big deal. Blobstore should expand to be a gateway to S3, etc.
- There are optimizations to be done for image storage and transfer. We can chunk and deduplicate with rabin fingerprinting (aka rolling checksum). This is used in rsync and git and is a very effective way to diff opaque binaries.
- I'm not convinced that we want full-on layered artifact support. The slugs we use are only the `/app` mountpoint, constructing filesystems based on mounts instead of overlays with file whiteouts etc, is easier.

**Comment by lmars** [comment]
@josephglanville thanks for the thoughts re OverlayFS / ZFS, I think it's a case of trying out the options and seeing what fits.

@titanous 

> I'm not convinced that we want full-on layered artifact support. The slugs we use are only the /app mountpoint, constructing filesystems based on mounts instead of overlays with file whiteouts etc, is easier.

The reason I like overlays over mounts is because they don't have to "own" an entire directory tree (e.g. if my artifact contains `/usr/lib/{xxx,yyy}` I can make just those files available without hiding the rest of `/usr/lib`).

This of course isn't an issue for slugs which do own the directory tree, but I am now wondering if it would be simpler to just have tar artifacts which just get extracted into the container's layer before it starts? For example:

```go
Artifact{
  Type:        ArtifactTypeTar,
  URI:         "http://blobstore.discoverd/xxx.tgz",
  Compression: "gzip",
  ExtractPath: "/app",
}
```

This is probably more obvious to the end user that subsequent tar artifacts when extracted will overwrite the same files from previous artifacts.

**Comment by titanous** [comment]
@lmars Yeah, tar artifacts sound like a good evolution from the current slug system. `ExtractPath` is ambiguous,  it could refer to a path inside the tar to extract or a target path to extract to.

**Comment by titanous** [comment]
In progress at #3100.

**Comment by lmars** [comment]
Implemented in #3619.
